### DISCUSSION TITLE
Looking for a new name for the avocado.utils namespace

### DISCUSSION THREAD

**Discussion by ana** [avocado-framework/avocado#5284] [RFC]
I'm working on a project to split `avocado.utils` and distribute it in a separate git repository, also with a different Python namespace.
This means we need a new name for this namespace!

There are mostly two options:
* Use something related: `afutils`, `avocadoutils`, `autils`, ...
* Use something totally related to avocado that could be easier to remember: `greenmatter`, `aguacate`,  `persea`, `abacado`,

What do you like? Do you have any other suggestions?

**Comment by clebergnu** [comment]
Hi Ana,

Thanks for bringing up this discussion!

My suggestion/vote is greenmatter.  Yes, I'm biased because that was a suggestion of mine.  The reasoning for the name is:

* (green)pulp is too obvious
* it will make a good part of what Avocado (as a project is), but it's not quite its core
* it somehow relates to "gray matter": yes, if you have some, you'd use it ;)
* it's available on PIP, Fedora packages, etc

**Reply by beraldoleal** [reply]
+1 for greenmatter.

**Comment by pevogam** [comment]
How about some more serious name that better explains the purpose of the library? I find all of the names above unclear and ambiguous and this is not a main product but a secondary repo with more functional purpose.

**Reply by clebergnu** [reply]
Hi @pevogam, do you have a name suggestion that would illustrate the (good) points you make?

**Reply by pevogam** [reply]
How about something as simple and straightforward as `avocado-utils`? ^^

**Reply by richtja** [reply]
I agree with @pevogam that the name should better explain the purpose of the library. But I prefer shorter names over longer to save typing. So I would vote for `autils`.

**Reply by beraldoleal** [reply]
@pevogam, I would prefer avocado-utils too, however from PEP8:

> Modules should have short, all-lowercase names. Underscores can be used in the module name if it improves readability. Python packages should also have short, all-lowercase names, although the use of underscores is discouraged.

Because of that I think we should avoid `avocado_utils`. `avocadoutils` would be the closest one, but I don't think it is "more serious" then "greematter". I guess is just a matter of taste.

**Reply by pevogam** [reply]
@beraldoleal, as @richtja suggested there are viable shorter options too like `autils` or `avoutils`.

But don't mind me too much, just sharing generic thoughts too.

**Reply by pevogam** [reply]
I sometimes find myself abbreviating avocado stuff using `avo` prefix.

**Reply by ana** [reply]
FWIW my preference is also with something short and related to the current name, so my preferences in order are: `afutils`, `autils`, `avutils`

**Reply by beraldoleal** [reply]
small note: `autils` is not available on pypi.

**Reply by ana** [reply]
> small note: `autils` is not available on pypi.

The name in pypi is indepdant, isn't?
Like it could be avocado-utils in pypi and later the modules be autils, greenmatter, etc

**Reply by beraldoleal** [reply]
They are, but I would prefer to have the same for packages as it is for the Python package, for consistency.

**Comment by lmr** [comment]
It's a utility library for test frameworks, any name would be fine, given
that any bystander can check what the name is about fairly quickly.

For example, one of the libraries used to write CLIs in golang is called
cobra [1]. It's not a very descriptive name in and of itself. But a simple
web search explains what it is about.

Given that I don't think this is a big concern and we can pick any type of
name, even names that don't immediately relate to its purpose. I tend to
prioritize shorter names over long ones to save some typing, but that's my
personal opinion.

That's just my 2 cents of course.

[1] https://github.com/spf13/cobra

On Fri, Mar 11, 2022 at 11:21 PM Plamen Dimitrov ***@***.***>
wrote:

> How about some more serious name that better explains the purpose of the
> library? I find all of the names above unclear and ambiguous and this is
> not a main product but a secondary repo with more functional purpose.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/avocado-framework/avocado/discussions/5284#discussioncomment-2342759>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACIOZ536N2DIJOTVYHGJFLU7PBM5ANCNFSM5QP65KKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: <avocado-framework/avocado/repo-discussions/5284/comments/2342759@
> github.com>
>


-- 
Lucas

**Reply by pevogam** [reply]
I agree that more freedom could be taken with a product name in comparison to module/function/class names within it, I just think there must be a tolerance border of how disconnected the product name must be from its original purpose. The best project names are usually formed in some clever ways that nicely integrate its semantics and purpose.

But I guess the choice of name can be quite subjective and while seeming very relevant for some of us it might seem quite irrelevant for others.

**Reply by clebergnu** [reply]
I believe the project name is already in a category where the name is as Lucas describes it: something that has no meaning in itself but that can refer people pretty quickly to it.

I find it odd to attempt to break from that in one repo of the project.  So, suppose we name it "avocado-utils", is "avocado" serious enough?  IMO, "avutils" attempts to bring some meaning in itself, but fails to do so.

I also agree with Lucas' point about short name.  The problem is that it's pretty hard to come by those short and somewhat unique names.  With that in mind, `persea` is my second choice because it follows the project naming "style", is unique and shorter than "greenmatter".

**Reply by pevogam** [reply]
I will quote what a colleague of mine would say for these matters: This is not a single, standalone product and if it were having an exotic name would be fine. However the sole reason for the repo to exist is to serve the avocado project, so in this context it has to make sense. It doesn't exist on its own, in that if the avocado project vanished it wouldn't be maintained anymore. Its name and purpose are therefore supportive from the very beginning.

**Reply by clebergnu** [reply]
About being a standalone product, the new utility repo should actually be standalone.  Such a repo would be related to the Avocado project (its origin, the libraries' main use cases) but there should be no requirement or coupling.  That's why I believe a "related" name makes sense.

If one is writing a tool that needs to interact with GDB and finds the `*.gdb` module to be helpful, there would be no need to have it run *under* Avocado.  The same goes for any other library.

Even if the Avocado test runner vanishes because a better alternative comes up, I believe there would still be valuable in the utility library, and it could survive past the Avocado test runner.

**Comment by ana** [comment]
Hi folks,

I propose we use the Condorcet method here, it's a voting system in which voters rank their options in a sequence of 1st, 2nd, 3rd, etc.
To get the results, I'll use http://www.ericgorr.net/condorcet/   (method Condorcet - Basic)

Voting will be done as a comment to this post with your ranking, I'll explain how this should be done later. 
The current list of options to vote are:

1 greenmatter (proposed @clebergnu)
2 avocado-utils (proposed by @pevogam)
3 autils (proposed by @richtja)
4 persea (proposed by @lmr)
5 afutils (proposed by @ana)

Please, tell me if I'm forgetting any here. @beraldoleal do you want to add a new one?

**Reply by beraldoleal** [reply]
Here is my order, and a suggestion to use the meso part of an avocado:

  1 greenmatter (proposed @clebergnu)
  2 mesocarp (proposed by @beraldoleal)
  3 persea (proposed by @lmr)
  4 autils (proposed by @richtja)
  5 afutils (proposed by @ana)
  6 avocado-utils (proposed by @pevogam)

**Reply by ana** [reply]
@beraldoleal Voting is still not open. If I understand well, you want to add "mesocarp" to the list of options to vote?

**Reply by pevogam** [reply]
I also suggest even shorter options like `avoutils` and `avutils` to the table if this is possible.

**Reply by ana** [reply]
> I also suggest even shorter options like `avoutils` and `avutils` to the table if this is possible.

Could please limit yourself maximum to two options? If not we'll have a huge list for voting...

**Reply by pevogam** [reply]
I am not sure how much sense all of this makes considering there might be more good options than the number of people participating in this. Not to mention some [shortcomings](https://en.wikipedia.org/wiki/Condorcet_paradox) of the current and pretty much any voting methods ^^

So then simply choose whichever of the three you like if you insist on this approach.

**Reply by ana** [reply]
@pevogam  Yeah, any voting system has issues. Condorcet was one simple enough to be understood quickly and allow to have several choices and ..... hopefully only one result :-)

**Comment by ana** [comment]
Let's vote, the options are:

> 1 greenmatter (proposed @clebergnu)
> 2 avocado-utils (proposed by @pevogam)
> 3 autils (proposed by @richtja)
> 4 persea (proposed by @lmr)
> 5 afutils (proposed by @ana)
> 6 mesocarp (proposed by @beraldoleal)
> 7 avutils (proposed by @pevogam)

Please, leave your vote in a comment to this parent comment before March 27th (included). If you don't want to vote, a comment is also welcome with this information.

Your vote should be one line following the format indicated at : http://www.ericgorr.net/condorcet/

A couple of examples to help:

Example A) 
3 = 5 = 7 > 2 > 1 > 4 > 6
Explanation: you don't care between autils/afutils/avutils,  your next choice is avocado-utils, and then the order is greenmatter, persea, and finally mesocarp

Example B)
1 = 4 = 6 > 3 = 5 = 7 > 2
Explanation: You don't care between greenmatter/persea/mesocarp, then between autils/afutils/avutils and then finally, you don't like at all avocado-utils.

**Reply by ana** [reply]
My vote is  `5 = 7 = 3 > 4 = 2> 6 >  1  `

**Reply by richtja** [reply]
My vote is `3 = 7 > 5  > 1 > 4 > 6 > 2 `

**Reply by lmr** [reply]
My vote is `4 = 6 > 1 > 7 = 3 = 5 = 2`

**Reply by MCWSea** [reply]
My vote is 4 = 1 = 2 = 3 > 5 = 6 =7

**Reply by beraldoleal** [reply]
`1 = 6 > 4 > 3 = 5 = 7`

**Reply by pevogam** [reply]
My vote is `7 = 3 > 5 > 2 > 4 > 6 > 1`

**Reply by clebergnu** [reply]
1 > 4 = 6 > 3 > 7 > 5

**Reply by ldoktor** [reply]
`3 > 2 > 4 = 6 > 5 = 7 > 1`  (We have aexpect, which is avocado's version of pexpect; I became "allergic" to -matter suffixes, av- means audio-video, af- is connected to network (AF_INET6, ...))

**Reply by clebergnu** [reply]
> `3 > 2 > 4 = 6 > 5 = 7 > 1` (We have aexpect, which is avocado's version of pexpect; I became "allergic" to -matter suffixes, av- means audio-video, af- is connected to network (AF_INET6, ...))

I agree with most of your points here, and I'm curious about the `-matter*` projects/repos you interacted with.

**Reply by ana** [reply]
Hi folks, 
With the current votes here, I get the following ranked ballots:
```
1: 5 = 7 = 3 > 4 = 2> 6 > 1
1: 3 = 7 > 5 > 1 > 4 > 6 > 2
1: 4 = 6 > 1 > 7 = 3 = 5 = 2
1: 4 = 1 = 2 = 3 > 5 = 6 =7
1: 1 = 6 > 4 > 3 = 5 = 7
1: 7 = 3 > 5 > 2 > 4 > 6 > 1
1: 1 > 4 = 6 > 3 > 7 > 5
1: 3 > 2 > 4 = 6 > 5 = 7 > 1 
```
The result in the [voting calculator ](http://www.ericgorr.net/condorcet/) gives option 3 `autils` as winner.

People still can vote until Sunday 27 (included).

**Comment by clebergnu** [comment]
I just occurred to me that previous contributors to Avocado (and especially to `avocado/utils`) may want to have a say here.  I can think of @ldoktor @apahim @wainersm .  Feel free to notify other contributors.

**Comment by dzhengfy** [comment]
@ana , libvirt QE voted as following, thanks.

3 people like avocado-utils by @chunfuwen @kylazhang @smitterl
2 people like avutils by @dzhengfy and @brucepatricklee
3 people like autils by @Yingshun and @nanli1 @cliping

**Reply by ana** [reply]
Thanks @dzhengfy 
If you don't want to provide a ranked ballot, I only can interpret these results as:

```
3: 2 > 1 = 3 = 4 = 5 = 6 = 7
2: 7 > 1 = 2 = 3 = 4 = 5 = 6
3: 5 > 1 = 2 = 3 = 4 = 6 = 7
```

Adding to the votes so far, I get:

```
1: 5 = 7 = 3 > 4 = 2> 6 > 1
1: 3 = 7 > 5 > 1 > 4 > 6 > 2
1: 4 = 6 > 1 > 7 = 3 = 5 = 2
1: 4 = 1 = 2 = 3 > 5 = 6 =7
1: 1 = 6 > 4 > 3 = 5 = 7
1: 7 = 3 > 5 > 2 > 4 > 6 > 1
1: 1 > 4 = 6 > 3 > 7 > 5
1: 3 > 2 > 4 = 6 > 5 = 7 > 1 
3: 2 > 1 = 3 = 4 = 5 = 6 = 7
2: 7 > 1 = 2 = 3 = 4 = 5 = 6
3: 5 > 1 = 2 = 3 = 4 = 6 = 7
```

And the provisional winner would be `autils`. Voting closes on March 27.

**Comment by ana** [comment]
Thanks everybody who suggested names and voted. The winner is `autils`!
